prototypes/ScheduledMergesTestQLS.hs

```diff
@@ -209,6 +209,13 @@ instance StateModel (Lockstep Model) where
     ADuplicate :: ModelVar Model (LSM RealWorld)
                -> Action (Lockstep Model) (LSM RealWorld)
 
+    -- | Without this, the prototype only completes tiering merges when the next
+    -- merging run on this level is created, so a level would never contain a
+    -- completed merge.
+    ASupply :: ModelVar Model (LSM RealWorld)
+            -> Int
+            -> Action (Lockstep Model) ()
```

Comment:
It's not clear to me why this is necessary. Shouldn't the prototype be able to maintain it's invariants without supplying credits explicitly? If we add this action, we're not letting the prototype "do its thing"

Reply:
Without this, the whole part of the invariant that checks a `CompletedMerge` present in a tiering level is dead code. Runs stay `Ongoing` until they are replaced by the next incoming merge.

Maybe that's something I could change in the prototype, related to https://github.com/IntersectMBO/lsm-tree/pull/310#discussion_r1696639389.

Reply:
Without `ASupply`, I can do

```diff
--- a/prototypes/ScheduledMerges.hs
+++ b/prototypes/ScheduledMerges.hs
@@ -247,6 +247,7 @@ invariant = go 1
             -- Note that tiering on the last level only occurs when there is
             -- a single level only.
             (_, CompletedMerge r, MergeLastLevel) -> do
+              assertST $ False
               assertST $ ln == 1
               assertST $ tieringRunSizeToLevel r <= ln+1
 
@@ -254,7 +255,8 @@ invariant = go 1
             -- level it is entering, but can also be one smaller (in which case
             -- it'll be held back and merged again) or one larger (because it
             -- includes a run that has been held back before).
-            (_, CompletedMerge r, MergeMidLevel) ->
+            (_, CompletedMerge r, MergeMidLevel) -> do
+              assertST $ False
               assertST $ tieringRunSizeToLevel r `elem` [ln-1, ln, ln+1]
```

And even increase the generator size a lot and don't get any failures:

```
prototype
  ScheduledMerges
    ScheduledMerges vs model: OK (88.31s)
      +++ OK, passed 100 tests.
      
      Action polarity (83563 in total):
      100.000% +
      
      Actions (83563 in total):
      33.441% +AInsert
      32.943% +ALookup
      16.808% +ADelete
       8.383% +ADump
       8.308% +ADuplicate
       0.118% +ANew
```

Reply:
Although arguably this just putting on a bandaid. The lockstep tests just don't reach proper coverage. For the *first* tiering merge to even be created, ~100 elements need to be inserted, which means a lot of actions are required before anything interesting happens.

Reply:
I'll remove `ASupply` and instead make sure the invariant gets checked inbetween supplying credits and doing the other work (e.g. requiring a merge to be completed), so it has a chance to look at every merge once it completes. The need for better generators remains, though.

Reply:
One could add an action that performs multiple inserts (instead of just 1). Would that hit the corner case?

Reply:
I tried it, but it didn't catch any of the incorrect invariants I found with the handwritten tests. My implementation was not ideal, though, it can probably be improved. I only return one of the inserted keys (so it gets added to the available variables), which means key re-use is less likely and it's harder to get a lot of compaction. Returning multiple keys from a single action turned out to be tricky, so I abandoned that.